src/firebase-social-provider.js

```diff
@@ -374,3 +347,180 @@ FirebaseSocialProvider.prototype.setupDetectDisconnect_ = function() {
+      this.addUserProfile_(profile);
+    }.bind(this)).catch(function(e) {
+      // There is a race condition where if Alice invites Bob, when Bob accepts
+      // the invite and adds himself to Alice's friend directory, Alice will
+      // detect him in her /friends directory before she is added to Bob's
+      // /friends directory - this results in her being temporarily unable to
+      // read his profile.  Hack around this by waiting 1 second for Bob's
+      // /friends directory to be updated.
+      // TODO: try to fix this so we don't need to retry
+      setTimeout(function() {
+        this.readFriendProfile_(friendUserId).then(function(profile) {
+          this.addUserProfile_(profile);
+        }.bind(this)).catch(function(e) {
+          this.logger.error('Error reading profile for user ' + friendUserId);
+        }.bind(this));
+      }.bind(this), 1000);
+    }.bind(this));
+  }.bind(this));
+};
+
+
+FirebaseSocialProvider.prototype.acceptUserInvitation = function(jsonString) {
+  return new Promise(function(F, R) {
+    var data;
+    try {
+      data = JSON.parse(jsonString);
+    } catch (e) {
+      return R('Unable to parse: ' + jsonString);
+    }
+
+    // Try to write to friends friendRequest folder
+    var myUserId = this.getUserId_();
+    var friendUserId = data.userId;
+    var permissionToken = data.permissionToken;
+
+    // Sanity check
+    if (friendUserId === myUserId) {
+      return R('acceptUserInvitation called for logged in user');
+    }
+
+    var networkPrefix = this.networkName_ + ':';
+    var receivedPermissionTokensRef = new Firebase(
+        this.allUsersUrl_ + '/' + networkPrefix + myUserId +
+        '/receivedPermissionTokens/' + permissionToken);
+    receivedPermissionTokensRef.set({received: true}, function(error) {
+      if (error) {
+        return R('error writing to receivedPermissionTokens');
+      }
+
+      var friendRequestUrl = this.allUsersUrl_ + '/' + networkPrefix +
+          friendUserId + '/friends/' + networkPrefix + myUserId +
+          '/inviteResponses/' + permissionToken;
+      var friendRequestRef = new Firebase(friendRequestUrl);
+      friendRequestRef.update({isFriend: true}, function(error) {
+        if (error) {
+          return R('Not permissioned to add friend');
+        }
+        F();
+        // Add a directory for the friend.  This will trigger an update
+        // to the callback in loadContacts_ which will get the friend's
+        // profile, start monitoring them and emit an onUserProfile
+        this.addFriendDirectory_(friendUserId);
+      }.bind(this));
+    }.bind(this));
+  }.bind(this));
+};
+
+
+FirebaseSocialProvider.prototype.inviteUser = function(ignoredStringParam) {
+  if (!this.loginState_) {
+    throw 'Error in FirebaseSocialProvider.inviteUser: not logged in';
+  }
+
+  // Generate a random permissionToken and write it to Firebase.
+  var permissionToken = Math.floor(Math.random() * 100000000000);
```

Comment:
I'm going to leave Math.random in this pull request, but entered #34 to investigate further